File: src/main/webapp/app/services/apiService.js
(function () {
    'use strict';

    angular.module('petclinicApp').service('apiService', ['$http', '$log', 'apiErrorHandlerService', function ($http, $log, apiErrorHandlerService) {

        /** CreateUser (secured)
         *
         * Api URL: /api/users?
         *"Request": CreateUserDto {
         *    "role": Int,
         *    "username": String,
         *    "password": String
         *}
         *
         *"Response": ReadUserResponseDto {
         *    "id": Int,
         *    "role": Int,
         *    "username": String
         *}
         */
        this.createUser = function (model, successCallback, errorCallback) {
            $http({
                method: 'POST',
                url: '/api/users',
                data: {
                  role: model.role,
                  username: model.username,
                  password: model.password
                }
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** ReadUser (secured)
         *
         * Api URL: /api/users/:id?
         *"Request": ReadUserDto {
         *    "id": Int
         *}
         *
         *"Response": ReadUserResponseDto {
         *    "id": Int,
         *    "role": Int,
         *    "username": String
         *}
         */
        this.readUser = function (model, successCallback, errorCallback) {
            $http({
                method: 'GET',
                url: '/api/users/' + model.id + ''
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** UpdateUser (secured)
         *
         * Api URL: /api/users/:id?
         *"Request": UpdateUserDto {
         *    "id": Int,
         *    "role": Int,
         *    "password": Option[String]
         *}
         *
         *"Response": ReadUserResponseDto {
         *    "id": Int,
         *    "role": Int,
         *    "username": String
         *}
         */
        this.updateUser = function (model, successCallback, errorCallback) {
            $http({
                method: 'PUT',
                url: '/api/users/' + model.id + '',
                data: {
                  role: model.role,
                  password: model.password
                }
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** DeleteUser (secured)
         *
         * Api URL: /api/users/:id?
         *"Request": ReadUserDto {
         *    "id": Int
         *}
         *"Response": Unit
         */
        this.deleteUser = function (model, successCallback, errorCallback) {
            $http({
                method: 'DELETE',
                url: '/api/users/' + model.id + ''
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** Users (secured)
         *
         * Api URL: /api/users?from:Int&maxRowCount:Int
         *"Request": UsersDto {
         *    "from": Int,
         *    "maxRowCount": Int
         *}
         *"Response": SearchResultDto[UsersResponseDto]
         */
        this.users = function (from, maxRowCount, model, successCallback, errorCallback) {
            $http({
                method: 'GET',
                url: '/api/users',
                params: {
                    from: from,
                    maxRowCount: maxRowCount
                }
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** AdminUsers (secured)
         *
         * Api URL: /api/adminUsers?role:Int&from:Int&maxRowCount:Int
         *"Request": AdminUsersDto {
         *    "role": Int,
         *    "from": Int,
         *    "maxRowCount": Int
         *}
         *"Response": SearchResultDto[AdminUsersResponseDto]
         */
        this.adminUsers = function (role, from, maxRowCount, model, successCallback, errorCallback) {
            $http({
                method: 'GET',
                url: '/api/adminUsers',
                params: {
                    role: role,
                    from: from,
                    maxRowCount: maxRowCount
                }
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** SignUp
         *
         * Api URL: /api/signUp?
         *"Request": SignUpDto {
         *    "username": String,
         *    "passwordHash": String
         *}
         *
         *"Response": AuthenticationResponseDto {
         *    "username": String,
         *    "roleId": Int,
         *    "authenticationCode": String
         *}
         */
        this.signUp = function (model, successCallback, errorCallback) {
            $http({
                method: 'POST',
                url: '/api/signUp',
                data: {
                  username: model.username,
                  passwordHash: model.passwordHash
                }
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** SignIn
         *
         * Api URL: /api/signIn?
         *"Request": SignInDto {
         *    "username": String,
         *    "passwordHash": String
         *}
         *
         *"Response": AuthenticationResponseDto {
         *    "username": String,
         *    "roleId": Int,
         *    "authenticationCode": String
         *}
         */
        this.signIn = function (model, successCallback, errorCallback) {
            $http({
                method: 'POST',
                url: '/api/signIn',
                data: {
                  username: model.username,
                  passwordHash: model.passwordHash
                }
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** SignOut (secured)
         *
         * Api URL: /api/signOut?
         *"Response": Unit
         */
        this.signOut = function (successCallback, errorCallback) {
            $http({
                method: 'POST',
                url: '/api/signOut'
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

        /** Authenticate
         *
         * Api URL: /api/authenticate?
         *"Request": AuthenticationCodeDto {
         *    "authenticationCode": String
         *}
         *
         *"Response": AuthenticationResponseDto {
         *    "username": String,
         *    "roleId": Int,
         *    "authenticationCode": String
         *}
         */
        this.authenticate = function (model, successCallback, errorCallback) {
            $http({
                method: 'POST',
                url: '/api/authenticate',
                data: {
                  authenticationCode: model.authenticationCode
                }
            }).success(function (data) {
                successCallback(data);
            }).error(function (data, header, status, config) {
                apiErrorHandlerService(data, header, status, config, errorCallback);
            });
        };

    }]);
})();

File: android/app/src/main/java/eu/execom/dry/petclinic/api/ApiService.java
package eu.execom.dry.petclinic.api;

import android.content.Context;
import android.text.TextUtils;
import com.android.volley.*;
import com.android.volley.toolbox.Volley;
import eu.execom.dry.petclinic.api.dto.*;
import eu.execom.dry.petclinic.api.enumeration.*;
import java.util.HashMap;
import java.util.Map;

class ApiService {

    private final String TAG = "";
    private final String SECURITY_TOKEN = "X-PETCLINIC-AUTH";
    private final String serverHost;
    private final Context context;
    private RequestQueue mRequestQueue;
    private String securityToken = null;

    public ApiService(String serverHost, Context context) {
        this.serverHost = serverHost;
        this.context = context;
    }

    public String setSecurityToken() {
        return securityToken;
    }

    public void createUser(CreateUserDto requestDto, Response.Listener<ReadUserResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "users";
        Map<String, String> headers = new HashMap<String, String>();
        headers.put(SECURITY_TOKEN, securityToken);

        post(url, headers, requestDto, ReadUserResponseDto.class, listener, errorListener, tag);
    }

    public void readUser(ReadUserDto requestDto, Response.Listener<ReadUserResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "users/" + requestDto.getId() + "";
        Map<String, String> headers = new HashMap<String, String>();
        headers.put(SECURITY_TOKEN, securityToken);

        get(url, headers, ReadUserResponseDto.class, listener, errorListener, tag);
    }

    public void updateUser(UpdateUserDto requestDto, Response.Listener<ReadUserResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "users/" + requestDto.getId() + "";
        Map<String, String> headers = new HashMap<String, String>();
        headers.put(SECURITY_TOKEN, securityToken);

        UpdateUserBodyDTO bodyDto = new UpdateUserBodyDTO(requestDto.getRole(), requestDto.getPassword());

        put(url, headers, bodyDto, ReadUserResponseDto.class, listener, errorListener, tag);
    }

    public void deleteUser(ReadUserDto requestDto, Response.Listener<Void> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "users/" + requestDto.getId() + "";
        Map<String, String> headers = new HashMap<String, String>();
        headers.put(SECURITY_TOKEN, securityToken);

        delete(url, headers, Void.class, listener, errorListener, tag);
    }

    public void users(UsersDto requestDto, Response.Listener<UsersResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "users";
        Map<String, String> headers = new HashMap<String, String>();
        headers.put(SECURITY_TOKEN, securityToken);

        get(url, headers, UsersResponseDto.class, listener, errorListener, tag);
    }

    public void adminUsers(AdminUsersDto requestDto, Response.Listener<AdminUsersResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "adminUsers";
        Map<String, String> headers = new HashMap<String, String>();
        headers.put(SECURITY_TOKEN, securityToken);

        get(url, headers, AdminUsersResponseDto.class, listener, errorListener, tag);
    }

    public void signUp(SignUpDto requestDto, Response.Listener<AuthenticationResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "signUp";

        post(url, null, requestDto, AuthenticationResponseDto.class, listener, errorListener, tag);
    }

    public void signIn(SignInDto requestDto, Response.Listener<AuthenticationResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "signIn";

        post(url, null, requestDto, AuthenticationResponseDto.class, listener, errorListener, tag);
    }

    public void signOut(Response.Listener<Void> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "signOut";
        Map<String, String> headers = new HashMap<String, String>();
        headers.put(SECURITY_TOKEN, securityToken);

        post(url, headers, null, Void.class, listener, errorListener, tag);
    }

    public void authenticate(AuthenticationCodeDto requestDto, Response.Listener<AuthenticationResponseDto> listener, Response.ErrorListener errorListener, String tag) {
        String url = serverHost + "/" + "authenticate";

        post(url, null, requestDto, AuthenticationResponseDto.class, listener, errorListener, tag);
    }

    public void cancelPendingRequests(String tag) {
        getRequestQueue().cancelAll(tag);
    }

    protected <T> void get(String url, Map<String, String> headers, Class<T> responseClass,
                           Response.Listener<T> listener, Response.ErrorListener errorListener, String tag) {
        addToRequestQueue(new GsonRequest<T>(Request.Method.GET, url, headers, null, responseClass, listener, errorListener), tag);
    }

    protected <T> void post(String url, Map<String, String> headers, Object request, Class<T> responseClass,
                            Response.Listener<T> listener, Response.ErrorListener errorListener, String tag) {
        addToRequestQueue(new GsonRequest<T>(Request.Method.POST, url, headers, request, responseClass, listener, errorListener), tag);
    }

    protected <T> void put(String url, Map<String, String> headers, Object request, Class<T> responseClass,
                           Response.Listener<T> listener, Response.ErrorListener errorListener, String tag) {
        addToRequestQueue(new GsonRequest<T>(Request.Method.PUT, url, headers, request, responseClass, listener, errorListener), tag);
    }

    protected <T> void delete(String url, Map<String, String> headers, Class<T> responseClass,
                              Response.Listener<T> listener, Response.ErrorListener errorListener, String tag) {
        addToRequestQueue(new GsonRequest<T>(Request.Method.DELETE, url, headers, null, responseClass, listener, errorListener), tag);
    }

    private RequestQueue getRequestQueue() {
        // lazy initialize the request queue, the queue instance will be
        // created when it is accessed for the first time
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(context);
        }

        return mRequestQueue;
    }

    protected <T> void addToRequestQueue(Request<T> req, String tag) {
        req.setTag(TextUtils.isEmpty(tag) ? TAG : tag);        // set the default tag if tag is empty

        VolleyLog.d("Adding request to queue: %s", req.getUrl());
        getRequestQueue().add(req);
    }

}
